### PR TITLE
Data fetching

### LINKED ISSUE #80: JS: data fetching + DOM manipulation
## Description
Using https://jsonplaceholder.typicode.com/ apis, implement a web app like this:
- The index page shows 10 posts with their title, body and author name.
- Each post has an update and a delete button.
- When the update button is clicked, a native prompt is opened with current body of the post as default value. User then can update the body and submit it, which calls an api to update the post. After the update is successful, the body text should be changed in the page. During the update process, a pending indicator should be shown, such as a progress bar, a "updating..." button, or something.
- Same applies for delete button, except that instead of opening a prompt, it opens a confirm.
- There is an "Add post" button. When the button is clicked, the user is redirected to another page with some inputs for title, body, etc. The user fills the inputs and then clicks a submit button, which does exactly the same thing as updating a post (e.g. api call and showing pending indicator, etc.) When the submission is successful, the user is redirected back to the main page.
- There is a search input in the page. When the user types something in the input and clicks a "search" button, a search is done, showing only filtered posts. This filtering should be done server-side, using api calls.

It doesn't matter how you style the web app.

## What to google
- APIs
- fetch
- promises
- `window.prompt`, `window.confirm`
- DOM manipulation
- `querySelector`, `appendChild`, `removeChild`, `createElement`, `createTextNode`, etc.

### PR DESCRIPTION
**Pull request by MahtaMirzaei**
Closes: #80

### CODE REVIEW

File: mahta/JS/JS data fetching/script.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  });
+
+function redirectToNewPostPage() {
+  window.location.href = "new-post.html";
+}
+
+function submitPost() {
+  const post = {
+    title: document.getElementById("postTitleInput").value,
+    body: document.getElementById("postBodyInput").value,
```
Comment (Mr-MRF-Dev) [inline mahta/JS/JS data fetching/script.js:105]: Add `userId` to `post`.

File: mahta/JS/JS data fetching/script.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  fetch("https://jsonplaceholder.typicode.com/posts")
+    .then((response) => response.json())
+    .then((data) => {
+      const postsContainer = document.getElementById("posts-container");
+      postsContainer.innerHTML = "";
+
+      data.forEach((post) => {
+        const postElement = createPostElement(post);
+        postsContainer.appendChild(postElement);
+      });
```
Comment (Mr-MRF-Dev) [inline mahta/JS/JS data fetching/script.js:164]: Please create 10 new posts.

File: mahta/JS/JS data fetching/script.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      updateButton.addEventListener("click", () => {
+        console.log(`Update post with ID ${post.id}`);
+
+        const updatedBody = prompt("Enter the updated body:", post.body);
+
+        if (updatedBody !== null) {
+          body.textContent = "Updating...";
+          fetch(`https://jsonplaceholder.typicode.com/posts/${post.id}`, {
+            method: "PATCH",
+            body: JSON.stringify({ body: updatedBody }),
```
Comment (Mr-MRF-Dev) [inline mahta/JS/JS data fetching/script.js:45]: An example for updating posts (body):
[link](https://jsonplaceholder.typicode.com/guide/)

```js
body: JSON.stringify({
    id: 1,
    title: 'foo',
    body: 'bar',
    userId: 1,
  }),  
```

**Review by Mr-MRF-Dev — CHANGES_REQUESTED** [review]

File: mahta/JS/JS data fetching/script.js (hunk 2 above)
Comment (MahtaMirzaei) [reply mahta/JS/JS data fetching/script.js:164]: ? I didn't get what you mean

Comment (Mr-MRF-Dev) [reply mahta/JS/JS data fetching/script.js:164]: When the `getAllPost()` function is called, it requests 100 posts from the server, but we only need 10.

> The index page shows 10 posts with their title, body, and author name.

the function is used to generate posts on the index page. isn't it? (Line: 129)

@MahtaMirzaei

**Comment by Mr-MRF-Dev** [comment]
Where are the new changes?

**Review by Mr-MRF-Dev — CHANGES_REQUESTED** [review]
Where are the new changes? @MahtaMirzaei

**Comment by MahtaMirzaei** [comment]
> Where are the new changes? @MahtaMirzaei

Wasn't pushed completly last time.

File: mahta/JS/JS data fetching/script.js (hunk 2 above)
Comment (MahtaMirzaei) [reply mahta/JS/JS data fetching/script.js:189]: I think it's solved.. but I'm not sure if I must change the part were i use slice for the code.

File: mahta/JS/JS data fetching/new-post.html [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Posts App</title>
+    <link rel="stylesheet" type="text/css" href="styles.css" />
+  </head>
+  <body>
+      <div class="add">
+      <input type="text" id="postTitleInput" placeholder="Title" />
+      <input type="text" id="postUserIdInput" placeholder="UserID" />
```
Comment (Mr-MRF-Dev) [inline mahta/JS/JS data fetching/new-post.html:10]: Rewrite the input tag and get the number from the client.

File: mahta/JS/JS data fetching/script.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
+function createPostElement() {
```
Comment (Mr-MRF-Dev) [inline mahta/JS/JS data fetching/script.js:1]: It is required to optimize the function to create an element tag efficiently without creating all post and body code.

**Review by Mr-MRF-Dev — CHANGES_REQUESTED** [review]

File: mahta/JS/JS data fetching/new-post.html (hunk 4 above)
Comment (MahtaMirzaei) [reply mahta/JS/JS data fetching/new-post.html:10]: I think it's already done in my submitPost function, but I'm waiting for further instructions if it's not what you meant.

Comment (Mr-MRF-Dev) [reply mahta/JS/JS data fetching/new-post.html:10]: I can submit a fake ID with characters like 'ABC' :)

Comment (Mr-MRF-Dev) [reply mahta/JS/JS data fetching/new-post.html:10]: @MahtaMirzaei

**Review by Mr-MRF-Dev — DISMISSED** [review]

**Review by mkermani144 — CHANGES_REQUESTED** [review]
It seems pending states are not respected. There is a short time between when the user confirms an update/delete and when (s)he sees the updates result or deletion. Meanwhile, a pending indicator (e.g. "updating post...", "deleting post...", etc.) should be shown.

**Review by mkermani144 — APPROVED** [review]

**Pull request merged**